src/lsm/grid.zig

```diff
@@ -91,49 +240,107 @@ pub fn GridType(comptime Storage: type) type {
             };
 
+            grid.start_read(read);
+        }
+
+        fn start_read(grid: *Grid, read: *Grid.Read) void {
+            // Check if a read is already in progress for the target address.
+            {
+                var it = grid.read_iops.iterate();
+                while (it.next()) |iop| {
+                    if (iop.reads.peek().?.address == read.address) {
+                        assert(iop.reads.peek().?.checksum == read.checksum);
+                        iop.reads.push(read);
+                        return;
+                    }
+                }
+            }
+
+            const iop = grid.read_iops.acquire() orelse {
+                grid.reads_pending.push(read);
+                return;
+            };
+
+            const block = grid.cache.put_no_clobber_preserve_locked(
+                *Grid,
+                block_locked,
+                grid,
+                read.address,
+            );
+
+            iop.* = .{
+                .grid = grid,
+                .completion = undefined,
+                .block = block,
+            };
+            iop.reads.push(read);
+
             grid.superblock.storage.read_sectors(
                 read_block_callback,
-                &read.completion,
-                block,
-                block_offset(address),
+                &iop.completion,
+                iop.block,
+                block_offset(read.address),
             );
         }
 
+        inline fn block_locked(grid: *Grid, block: BlockPtrConst) bool {
+            var it = grid.read_iops.iterate();
+            while (it.next()) |iop| {
+                if (block == iop.block) return true;
+            }
+            return false;
+        }
+
         fn read_block_callback(completion: *Storage.Read) void {
-            const read = @fieldParentPtr(Read, "completion", completion);
+            const iop = @fieldParentPtr(ReadIOP, "completion", completion);
 
-            const header_bytes = read.block[0..@sizeOf(vsr.Header)];
+            const header_bytes = iop.block[0..@sizeOf(vsr.Header)];
             const header = mem.bytesAsValue(vsr.Header, header_bytes);
-            const body = read.block[@sizeOf(vsr.Header)..header.size];
+            const body = iop.block[@sizeOf(vsr.Header)..header.size];
+
+            const address = iop.reads.peek().?.address;
+            const checksum = iop.reads.peek().?.checksum;
 
-            if (header.checksum == read.checksum and
+            if (header.checksum == checksum and
                 header.valid_checksum() and
                 header.valid_checksum_body(body))
             {
-                read.finish();
+                while (iop.reads.pop()) |read| {
+                    assert(read.address == address);
+                    assert(read.checksum == checksum);
+                    read.finish(iop.block);
+                }
             } else {
-                read.grid.read_recovery_queue.push(read);
```

Comment:
It would be good to error log which of the 3 reasons (invalid checksum or body or checksum mismatch) meant that we had to push to the recovery queue.